stdlib/public/runtime/Casting.cpp

```diff
@@ -2876,10 +2876,40 @@ id _bridgeAnythingNonVerbatimToObjectiveC(OpaqueValue *src,
 extern "C" const _ObjectiveCBridgeableWitnessTable BRIDGING_CONFORMANCE_SYM;
 #endif
 
+/// Nominal type descriptor for Swift.String.
+extern "C" const StructDescriptor NOMINAL_TYPE_DESCR_SYM(SS);
+
+static const _ObjectiveCBridgeableWitnessTable *
+swift_conformsToObjectiveCBridgeable(const Metadata *T) {
+  return reinterpret_cast<const _ObjectiveCBridgeableWitnessTable *>
+    (swift_conformsToProtocol(T, &PROTOCOL_DESCR_SYM(s21_ObjectiveCBridgeable)));
+}
+
 static const _ObjectiveCBridgeableWitnessTable *
 findBridgeWitness(const Metadata *T) {
-  auto w = swift_conformsToProtocol(T,
-                                &PROTOCOL_DESCR_SYM(s21_ObjectiveCBridgeable));
+  // Special case: Memoize the bridge witness for Swift.String.
+  // Swift.String is the most heavily used bridge because of the prevalence of
+  // string-keyed dictionaries in Obj-C.  It's worth burning a few words of static
+  // storage to avoid repeatedly looking up this conformance.
+  if (T->getKind() == MetadataKind::Struct) {
+    auto structDescription = cast<StructMetadata>(T)->Description;
+    // Check for a _very_ small number of heavily-used types.
+    // Cache the protocol witness for these to avoid lookup overhead.
+    if (structDescription == &NOMINAL_TYPE_DESCR_SYM(SS)) {
+      static auto *Swift_String_ObjectiveCBridgeable = swift_conformsToObjectiveCBridgeable(T);
```

Comment:
I think for this pattern to work, one needs to have an initializer. I don't think it works with pointers. So one would need a struct that does this in the constructor.

Reply:
I take it back! Try this on god bolt:

```

int myGlobal[5] = {0,1,2,3,4};

struct Foo {
    int *ptr;
    Foo(int offset) {
        ptr = &myGlobal[offset];
    }
};

extern int *getGlobal(int offset) {
    static int *ptr = &myGlobal[offset];
    return ptr;
}
```

I am seeing this with clang-9:

```
getGlobal(int):                          # @getGlobal(int)
        push    rbx
        mov     al, byte ptr [rip + guard variable for getGlobal(int)::ptr]
        test    al, al
        je      .LBB0_1
.LBB0_3:
        mov     rax, qword ptr [rip + getGlobal(int)::ptr]
        pop     rbx
        ret
.LBB0_1:
        mov     ebx, edi
        mov     edi, offset guard variable for getGlobal(int)::ptr
        call    __cxa_guard_acquire
        test    eax, eax
        je      .LBB0_3
        movsxd  rax, ebx
        lea     rax, [4*rax + myGlobal]
        mov     qword ptr [rip + getGlobal(int)::ptr], rax
        mov     edi, offset guard variable for getGlobal(int)::ptr
        call    __cxa_guard_release
        jmp     .LBB0_3
```

Reply:
The reference you pointed out earlier describes single static initialization for any variable with static scope.  As I read it, it has nothing to do with whether there is a class involved.  Disassembling the function in question shows that the static variable here does indeed have its initialization wrapped in two checks:  an outer check for the variable being null (uninitialized) and an inner block that is guarded by `__cxa_guard_acquire` and `__cxa_guard_release`.  As for performance, the benchmarks above show no substantial difference, so I think it's best to stick with the standard portable mechanism.